clippy_lints/src/utils/mod.rs

```diff
@@ -50,17 +50,77 @@ use rustc_lint::{LateContext, Level, Lint, LintContext};
 use rustc_middle::hir::map::Map;
 use rustc_middle::ty::subst::{GenericArg, GenericArgKind};
 use rustc_middle::ty::{self, layout::IntegerExt, Ty, TyCtxt, TypeFoldable};
+use rustc_session::Session;
 use rustc_span::hygiene::{ExpnKind, MacroKind};
 use rustc_span::source_map::original_sp;
 use rustc_span::sym as rustc_sym;
 use rustc_span::symbol::{self, kw, Symbol};
 use rustc_span::{BytePos, Pos, Span, DUMMY_SP};
 use rustc_target::abi::Integer;
 use rustc_trait_selection::traits::query::normalize::AtExt;
+use semver::Version;
 use smallvec::SmallVec;
 
 use crate::consts::{constant, Constant};
 
+pub fn parse_msrv(msrv: Option<String>, sess: Option<&Session>, span: Option<Span>) -> Option<Version> {
+    match msrv {
+        Some(val) => match Version::parse(&val) {
+            Ok(version) => Some(version),
+            Err(_) => {
+                if sess.is_some() && span.is_some() {
+                    sess.unwrap()
+                        .span_err(span.unwrap(), &format!("{} is not a valid Rust version!", val));
+                }
+                None
+            },
+        },
+        _ => None,
+    }
+}
+
+pub fn meets_msrv(msrv: Option<&Version>, lint_msrv: &str) -> bool {
+    if let Some(msrv) = msrv {
+        msrv >= &Version::parse(lint_msrv).unwrap()
+    } else {
+        true
+    }
+}
+
+#[macro_export]
+macro_rules! extract_msrv_attr {
+    (LateContext) => {
+        fn enter_lint_attrs(&mut self, cx: &rustc_lint::LateContext<'tcx>, attrs: &'tcx [Attribute]) {
+            match get_inner_attr(cx.sess(), attrs, "msrv") {
+                Some(msrv_attr) => {
+                    if let Some(msrv) = msrv_attr.value_str() {
+                        self.msrv =
+                            crate::utils::parse_msrv(Some(msrv.to_string()), Some(cx.sess()), Some(msrv_attr.span));
+                    } else {
+                        cx.sess().span_err(msrv_attr.span, "bad clippy attribute");
+                    }
+                },
+                _ => (),
+            }
+        }
+    };
+    (EarlyContext) => {
+        fn enter_lint_attrs(&mut self, cx: &rustc_lint::EarlyContext<'tcx>, attrs: &'tcx [Attribute]) {
+            match get_inner_attr(cx.sess, attrs, "msrv") {
+                Some(msrv_attr) => {
+                    if let Some(msrv) = msrv_attr.value_str() {
+                        self.msrv =
+                            crate::utils::parse_msrv(Some(msrv.to_string()), Some(cx.sess), Some(msrv_attr.span));
+                    } else {
+                        cx.sess.span_err(msrv_attr.span, "bad clippy attribute");
+                    }
+                },
+                _ => (),
+            }
+        }
+    };
+}
+
```

Comment:
@flip1995 I've put the `enter_lint_attrs` functions in a macro here but wasn't able to make it generic across `EarlyContext` and `LateContext` as the context type couldn't be inferred if `LintContext` was used as a parameter. Let me know if there's a better way to write this

Reply:
The solution with passing `{Early,Late}Context` to the macro looks good to me :+1: